Pixly

zed9h opened the pull request:
First try at implementing the Pixly file format, just loading for now and I tested with only two files, but it could be useful to some people ([for instance this guy](https://trello.com/c/lyMh9uSc/13-import-psd-aseprite-gif-pcx)), so I am pushing right away.

dacap commented:
Hi @zed9h, could this use the same png encoder/decoder to avoid duplicating the whole code? I see that it opens an XML file and a png file, so to load the png file you might try to use the png decoder (or calling `load_document()` to load the png file as a temporal `doc::Sprite`/`app::Document`).

Also could you please [sign our CLA](https://github.com/aseprite/opensource/blob/master/sign-cla.md#sign-the-cla)?

zed9h commented:
@dacap, maybe It would be cleaner if I created a helper to decode/encode the PNG for both file formats, but first I will work on the Pixly format save, which will add more value upfront.

PS: I will read the CLA.

dacap commented:
I'll take a look at this on next Tuesday. At the moment I recommend you to use `load_document()` if it's possible, so we can avoid modifying the actual png encoder/decoder (the code review from my side would be too much work if you refactor the code).

You can use `Document* tmpDoc = load_document(nullptr, pngfilename)` to "subload" the png file, and `save_document(nullptr, tmpDoc)` to save the png file.

zed9h commented:
I signed the CLA and finished the Pixly file save; the code ended up much simpler than the original PNG format, because it only handle RGBA, take a look if you think it worth the generalization of the codec. For me it was easier to handle a standard lib code than aseprite internal structure, even when I found the latter pretty clear and organized.

dacap commented:
@zed9h Remember that there is [an email required](https://github.com/aseprite/opensource/blob/master/sign-cla.md#sending-the-email) in the CLA signing process too.

dacap commented on src/app/file/pixly_format.cpp:
```diff
@@ -0,0 +1,487 @@
+    "<PixlyAnimation version=\"1.5\">\n"
+    "\t<Info "
+    "sheetWidth=\"%d\" sheetHeight=\"%d\" "
+    "totalCollumns=\"%d\" totalRows=\"%d\" "
```
Is this `totalCollumns` or `totalColumns`?

dacap commented on src/app/file/pixly_format.cpp:
```diff
@@ -0,0 +1,487 @@
+      sprite->folder()->addLayer(new LayerImage(sprite));
+    }
+
+    int *visible = (int*)malloc(layerCount * sizeof(int));
```
`visible` can be a `std::vector`, e.g.

    std::vector<int> visible(layerCount, 0);

dacap commented on src/app/file/pixly_format.cpp:
```diff
@@ -0,0 +1,487 @@
+// Aseprite
+// Copyright (C) 2001-2016  David Capello
```
You can add a new Copyright (C) 2016 [your name] here.

dacap commented on src/app/file/pixly_format.cpp:
```diff
@@ -0,0 +1,487 @@
+    TiXmlHandle xml(doc.get());
+    fop->setProgress(0.75);
+
+    TiXmlElement* xmlAnim = xml.FirstChild("PixlyAnimation").ToElement();
```
What happen if `PixlyAnimation` element does not exist?

dacap commented on src/app/file/pixly_format.cpp:
```diff
@@ -0,0 +1,487 @@
+
+    TiXmlElement* xmlAnim = xml.FirstChild("PixlyAnimation").ToElement();
+
+    TiXmlElement* xmlInfo = xmlAnim->FirstChild("Info")->ToElement();
```
Same here, what happen if `Info` element does not exist? (Or the attributes in the following lines, `layerCount`, `frameWidth`, `frameHeight`?)

dacap commented on src/app/file/pixly_format.cpp:
```diff
@@ -0,0 +1,487 @@
+
+    TiXmlElement* xmlInfo = xmlAnim->FirstChild("Info")->ToElement();
+
+    int layerCount  = strtol(xmlInfo->Attribute("layerCount"), NULL, 10);
```
You should use `base::convert_to<int>(...)` to convert from `std::string`/`const char*` to `int`

dacap commented on src/app/file/pixly_format.cpp:
```diff
@@ -0,0 +1,487 @@
+
+      const char * visible_str = xmlFrame->Attribute("visible");
+      if(visible_str) {
+        visible[(int)layer_index] += std::string(visible_str) == "true";
```
Add `(`and `)` in the right-side expression:

    visible[(int)layer_index] += (std::string(visible_str) == "true");

dacap commented:
As a side note, I'll need several `.anim` files to test these routines (original files generated with Pixly), because I don't have an Android device and cannot install Pixly. You can send me some test files at support@aseprite.org (I'll add them into an internal test suite I use for decoders/encoders)

zed9h commented:
@dacap Great points, I placed many more check()s and improved the error messages. I will send some good files and some bad ones to test the exceptions.

dacap commented on src/app/file/pixly_format.cpp:
```diff
@@ -0,0 +1,535 @@
+{
+  const Sprite* sprite = fop->document()->sprite();
+
+  if(sprite->pixelFormat() != IMAGE_RGB || !fop->document()->sprite()->needAlpha()) {
```
There is no need to check `needAlpha`. Also, you can remove the `pixelFormat()` check if you remove the `FILE_SUPPORT_GRAY `, `FILE_SUPPORT_GRAYA`, and `FILE_SUPPORT_INDEXED` from `onGetFlags()` (and you should remove those constants from `onGetFlags()` as this format doesn't support Grayscale, Indexed)

dacap commented on src/app/file/pixly_format.cpp:
```diff
@@ -0,0 +1,535 @@
+      FILE_SUPPORT_INDEXED |
+      FILE_SUPPORT_LAYERS |
+      FILE_SUPPORT_FRAMES |
+      FILE_SUPPORT_PALETTE_WITH_ALPHA;
```
Remove `FILE_SUPPORT_PALETTE_WITH_ALPHA` too as this format doesn't support indexed images with color palettes (with alpha or without alpha).

dacap commented on src/app/file/pixly_format.cpp:
```diff
@@ -0,0 +1,535 @@
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
```
As a minor detail, you can `#include <cstdio>`, `<cstdlib>`, and `<cmath>` here.

zed9h commented:
If I remove the FILE_SUPPORT_PALETTE_WITH_ALPHA and the auto-generated palette has some alpha on it, there will be a warning every time the file is saved, so I think I will need to keep this flag, even if the palette will not be saved. I also added FILE_SUPPORT_BIG_PALETTES for the same reason. (Not really sure if this is the right way to go.)

I cleaned up some includes that seemed unnecessary.

dacap commented:
Mhh, I think it's a bug in the code that check conditions in `file.cpp`. I'll be working in a fix for this so `FILE*PALETTE*` flags are checked only for Indexed images. (But I've to think more about those situations.)

The `FILE_SUPPORT_BIG_PALETTES` is in case the file format can save a color palette with more than 256 colors.

You can leave the flags and I'll then modify when the fix is done. Don't worry about that.

dacap commented on src/app/file/pixly_format.cpp:
```diff
@@ -0,0 +1,525 @@
+    int imageCount = check_number<int>(xmlFrames->Attribute("length"));
+
+    if(layerCount <= 0 || imageCount <= 0) {
+      throw Exception("No cels found");
```
I see that these `throw` can generate some leaks (mainly the png structures that are not freed). Anyway, don't worry, I'll integrate this with the png encoder/decoder later. I'll merge this code as it is.

dacap replied:
Oh sorry, I've just saw the try{}catch block to avoid leaks.

dacap commented:
Finally I got a crash [saving this file](https://github.com/aseprite/aseprite/files/346562/crash.ase.zip):

![screen shot 2016-07-04 at 12 34 09 pm](https://cloud.githubusercontent.com/assets/39654/16565455/a9b1e20a-41e3-11e6-9074-362015b3e8f7.png)

Could you please check this issue?

zed9h commented:
As I suspected when I saw the line of the error, the cel image can be smaller than the sprite size, I will compensate for that.

dacap commented:
Returning to this one: I'll finally remove `FILE_SUPPORT_BIG_PALETTES` and `FILE_SUPPORT_PALETTE_WITH_ALPHA` flags from `.anim` format as this format doesn't support these features (and the user must be warned that the palette will be lost). I've added the new issue #1182 how to fix the issues commented here.

zed9h commented:
Of course, if the palette is user created there should be a warning.

@dacap BTW, I didn't explain it properly here, the "collums" (sic) are on the official Pixly format, nothing I could do about it.

dacap commented:
> BTW, I didn't explain it properly here, the "collums" (sic) are on the official Pixly format, nothing I could do about it

👍   yeah, that's what I thought
